docs/architecture/adr-020-protobuf-transaction-encoding.md

```diff
@@ -54,26 +55,40 @@ which will be re-used by `SignDoc` below, and `signatures`:
 package cosmos_sdk.v1;
 
 message Tx {
-    TxBody body = 1;
-    AuthInfo auth_info = 2;
+    // A protobuf serialization of a TxBody that matches representation in SignDoc.
```

Comment:
Why does a `Tx` now have a `bytes` blob? Can we still achieve the desired outcome without this change? When I look at a tx, I would imagine I could understand and read all the fields.

Reply:
I tried to explain that in point 2. of the PR description, may be that helps. Anyways, let me try to go one step back to explain the underlying challenge. Maybe it is important to note that I'm not trying to change things here but cleanup the document based on what we agreed on.

A protobuf document is an abtract data structure that has a practically unlimited number of valid binary representation. Having a deterministic binary representation has never been a design goal of protobuf (like JSON and in contrast to other formarts like e.g. [Borsh](https://borsh.io/)). In order to verify signatures, the transaction signer and the verifier need to be able to agree on the exact same binary representation for the data. There are two ways to get there:
1. Specify a canonical representation (which is a custom addition to protobuf), that all implementations need to implement
2. Encode the protobuf document to bytes only once and maintain the binary representation

There is nothing wrong with 1. or 2., they just have different trade-offs. Aaron initially worked towards 1., but during the course of development explicitely discarded the idea of canonicalization which is documented in this file.

So we're going for 2., i.e. the signer chooses one binary representation and signs exactly that. But how is this binary representation transmitted to the verifier that only gets a `Tx`? An abspract data model (`Tx` with embedded `TxBody`) is not sufficient as it does not define the exact bytes needed for verification.

--------

Since the messages are [embedded as binary data because of the Any type](https://github.com/protocolbuffers/protobuf/blob/v3.12.3/src/google/protobuf/any.proto#L153-L154), you need a multi step decoding anyways to transform the binary in a complete fully nested tx model in memory. There are 3 main decoding steps to get there:
1. Decode raw data into `Tx`
2. Decode a binary body into `TxBody`
3. For each message decode the bytes in the `Any` value into the message type

This three step decoding is not the most simple thing on earth, but given the requirements that were expressed along the way, I think it has aceptable complexity and is very clean.

Reply:
> But how is this binary representation transmitted to the verifier that only gets a Tx?

Same way it does it now. The verifier (i.e AnteHandler), computes the raw binary data and verified the sig. It takes what it needs to from the Tx and from the state (nonce and sequence) and verifies against the signature.

Reply:
> The verifier (i.e AnteHandler), computes the raw binary data and verified the sig. It takes what it needs to from the Tx and from the state (nonce and sequence) and verifies against the signature.

So this is implemented already? Could you please point me to the code you are referring to? I'm very unfamiliar with the codebase and was not very successful with full text search.

Reply:
https://github.com/cosmos/cosmos-sdk/blob/master/x/auth/ante/sigverify.go#L193-L214
https://github.com/cosmos/cosmos-sdk/blob/master/x/auth/types/stdtx.go#L248-L260
https://github.com/cosmos/cosmos-sdk/blob/master/x/auth/types/stdtx.go#L292-L313

Reply:
Thank you, this is very helpful. The last link (the implementation of `StdSignBytes`) shows that this implements the legazy JSON verification mode, not the direct protobuf verification. In legazy JSON there is a canonicalization step [sdk.MustSortJSON(bz), implemented here](https://github.com/cosmos/cosmos-sdk/blob/bd90d6b34448775e77b21676597c9fb42b1a65d3/types/utils.go#L24-L50).

In our discussions and also in noted in some places in ADR-020 it was decided to not require canonicalization of any protobuf struture other than the outermost `SignDoc`. SignDoc is simple enough to not require canonicalization other than _no extra data_ and _fields sorted by number_, which we probably get for free in all implementations.

We can challenge this decision, but we should be explicit about it.